nrfx_usbd: Copy nrfx usbd code to Zephyr

tmon-nordic opened the pull request:
Copy nrfx_usbd code from zephyrproject-rtos/hal_nordic git revision d054a315eb888ba70e09e5f6decd4097b0276d1f. This enables us to refactor the code towards a native driver in a step by step manner where the smallest changes that bring biggest performance improvements are done early on.

tmon-nordic committed:
nrfx_usbd: Copy nrfx usbd code to Zephyr

Copy nrfx_usbd code from zephyrproject-rtos/hal_nordic git revision
d054a315eb888ba70e09e5f6decd4097b0276d1f. This enables us to refactor
the code towards a native driver in a step by step manner where the
smallest changes that bring biggest performance improvements are done
early on.

The code is not a vanilla copy from zephyrproject-rtos/hal_nordic.
The code was reformated with clang-format to match project style.
Manual modifications were done to change comments formatting, place
constant comparisons on the right and add blank lines to pass Zephyr
compliance check.

Relicense to Apache 2.0.

No functional changes.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
nrf_usbd_common: Guard DMA with semaphore

Rely on semaphore to serialize access to DMA instead of busy looping
after triggering DMA. With this change Ozone Code Profile generated with
J-Trace Pro on nrf52840dk_nrf52840 board running headphones microphone
sample shows following Load changes (trace data was reset once playback
and recording started and percentages were taken when memcpy reached
200 000 Run Count):
  * usbd_dmareq_process() from 17.16% to 2.24%
  * memcpy() from 9.37% to 8.36%
  * nrf_usbd_common_irq_handler() from 8.89% to 10.88%

Mark nrf_usbd_common_stop() as static because the caller must acquire
DMA semaphore before calling this function and the only place where it
is used is already acquiring the semaphore.

Disable EP0 SETUP interrupt when there is active DMA on EP0 to eliminate
the need for aborting DMA on EP0. This code path should not really
happen in real life though because hosts must not issue new SETUP before
a relatively long timeout (at least 50 ms).

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
nrf_usbd_common: Remove early DMA process handling

Early DMA process handling actually decreases performance in real world
scenarios because real world scenarios tend to be CPU bound. Moreover
the actual optimization is questionable because DMA semaphore can only
be released after the respective endpoint end event is handled. Remove
early DMA processing altogether and only check pending DMA at the end of
interrupt handler.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
nrfx_usbd: Rename to nrf_usbd_common

Rename local usbd copy from nrfx_usbd to nrf_usbd_common and use it in
both USB stacks. Renaming header to nrf_usbd_common.h allows breaking
changes in exposed interface. Mark all doxygen comments as internal
because local usbd copy should not be treated as public interface
because we are under refactoring process that aims to arrive at native
driver and therefore drop nrf_usbd_common in the future.

Use Zephyr constructs directly instead of nrfx glue macros.

No functional changes.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
nrf_usbd_common: Disable started interrupt handling

There is no point in calling an empty function on every single DMA
transfer start. While the empty function is just BX LR the fact that is
is called on every DMA transfer makes its impact visible. This change
improves CDC ACM echo throughput by approximately 2%.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
nrf_usbd_common: Refactor interrupt handling code

Ensure strict order within interrupt processing to eliminate workaround
in IN transfer acknowledged handler. EPDATASTATUS and DMA done events
are processed in a way that eliminates the possibility for race condions
between interrupt handler and host IN tokens.

Store last started DMA endpoint and use it in common handler for all DMA
finished events. Unify use of ep dma waiting variable because atomics
only make sense if all accesses are through atomic functions - here the
accesses are guarded with critical section.

Do not disable SETUP interrupt when DMA transfers data on endpoint 0 but
simply postpone handling if currently active DMA is on endpoint 0.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

carlescufi commented on modules/hal_nordic/nrfx/nrfx_usbd.c:
```diff
@@ -0,0 +1,1876 @@
+ * Copyright (c) 2016 - 2023, Nordic Semiconductor ASA
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
```
Nordic owns this code, please re-release it under the Apache v2 license.

carlescufi reviewed (changes requested):
Please fix the style issues in the former nrfx code. Code imported into the zephyr tree must comply with the coding style

carlescufi commented on modules/hal_nordic/nrfx/nrfx_usbd.c:
```diff
@@ -0,0 +1,1892 @@
+/*
```
Is there a reason why you would just not fold this directly into https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/device/usb_dc_nrfx.c? What's the point of keeping two separate files?

tmon-nordic replied:
This is also used by the new stack (drivers/usb/udc/udc_nrf.c). The changes done here benefit for the old and new stack. I prefer to first simplify the code and only then do the merge.

carlescufi replied:
OK, makes sense.

tmon-nordic commented on drivers/usb/common/nrf_usbd_common/nrf_usbd_common.c:
```diff
@@ -1494,72 +1217,74 @@ static void usbd_enable(void)
+	/* Clear EPDATA event and only then get and clear EPDATASTATUS to make
+	 * sure we don't miss any event.
+	 */
+	if (NRF_USBD->EVENTS_EPDATA) {
+		NRF_USBD->EVENTS_EPDATA = 0;
+		epdatastatus = NRF_USBD->EPDATASTATUS;
+		NRF_USBD->EPDATASTATUS = epdatastatus;
```
In my opinion, the code is much easier to follow if it can be easily be matched against the Product Specification document. When we arrive at native driver I would like to not use any USBD HAL at all.

anangl replied:
There are certain advantages of using the nrf HAL, like the possibility of handling in that layer some hardware anomalies or differences in registers layout between SoC series. It also makes it easy to use nrfx drivers in simulation. I agree that in the case of USBD, those advantages might not be that evident at the moment, but in my opinion, we should stay consistent in using nrf HAL in nRF drivers. Also because this file already uses USBD HAL functions, I think we should keep consistency in this regard.

tmon-nordic replied:
What is evident is that USBD HAL is really complicating the understanding of what the driver does. In my view the "file already uses" is really "file still uses" because I see no point in changing something without a real cause.

If the real cause would be "be consistent regarding the use of USBD HAL", then my strongly preferred solution to that would be to not use USBD HAL at all.

anangl replied:
I don't quite understand why is that a problem to use the USBD HAL in those few added accesses to USBD registers. It's much less effort than removing all uses of that HAL. But I still think this file should be consistent in using or not using the USBD HAL, so please do what you prefer to achieve this.

tmon-nordic replied:
Insisting on being consistent in using or not using USBD HAL kinda defeats the "enables us to refactor the code towards a native driver in a step by step manner". As you noticed it would be definitely harder to remove all remaining uses of USBD HAL at this point and the removal itself at this point serves no rel purpose but to please the reviewer.

If you still insist, then fine, I'll remove all uses of USBD HAL but I don't think it is worth the hassle at this point. And at the same time sticking to using the USBD HAL in newly rewritten code just for [foolish consistency](https://peps.python.org/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds) is silly. If you insist then it'll be next commit and the commit message will include that it is done without purpose other than to please the reviewer.

tmon-nordic commented on drivers/usb/common/nrf_usbd_common/nrf_usbd_common.c:
```diff
@@ -1687,6 +1706,9 @@ void nrf_usbd_common_disable(void)
 	nrf_usbd_common_stop();
 
+	/* Make sure DMA is not active */
+	k_sem_take(&dma_available, K_FOREVER);
```
I messed up during the rebases (when doing prefix change) and the code isn't really matching the commit message where I wrote `Mark nrf_usbd_common_stop() as static because the caller must acquire DMA semaphore before calling this function and the only place where it is used is already acquiring the semaphore.`. I will change the code to match the comment - i.e. take semaphore before calling `nrf_usbd_common_stop()`.

tmon-nordic commented on drivers/usb/common/nrf_usbd_common/nrf_usbd_common.c:
```diff
@@ -235,14 +236,11 @@ static uint32_t m_ep_ready;
  * cannot be started when there is ongoing transfer on any other channel.
  */
-static bool m_dma_pending;
+K_SEM_DEFINE(dma_available, 1, 1);
```
Fixed.

tmon-nordic commented on drivers/usb/common/nrf_usbd_common/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+zephyr_include_directories(.)
+
+zephyr_library_sources_ifdef(CONFIG_NRF_USBD_COMMON    nrf_usbd_common.c)
```
Is it ok now?

tmon-nordic commented on drivers/usb/common/nrf_usbd_common/nrf_usbd_common.c:
```diff
@@ -1286,7 +1309,7 @@ static inline size_t usbd_ep_iso_capacity(nrf_usbd_common_ep_t ep)
 static void usbd_dmareq_process(void)
 {
-	if (!m_dma_pending) {
+	if (m_ep_dma_waiting & m_ep_ready && !k_sem_take(&dma_available, K_NO_WAIT)) {
```
Changed.

tmon-nordic committed:
nrf_usbd_common: Keep interrupt configuration constant

Do not enable nor disable endpoint done interrupts at runtime because it
is not necessary. Simply keep all relevant interrupts enabled when USBD
is active and disable all interrupts when USBD is disabled.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
nrf_usbd_common: Remove dynamic feeders/consumers

There is no point in setting the feeder or consumer at runtime. The time
saved due to not checking conditions on each transaction is smaller than
the additional cost of calling function via pointer.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
nrf_usbd_common: Do not use USBD HAL

Use MDK directly instead of via USBD HAL to not mix direct register
accesses with USBD HAL. In my opinion this change is not really
necessary but reviewers are strict about consistency.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>
